app/controllers/concerns/rate_limit_concern.rb

```diff
@@ -37,26 +41,33 @@ def rate_limited_redirect(throttle_type)
     when :idv_doc_auth
       redirect_to idv_session_errors_throttled_url
     when :proof_address
-      redirect_to idv_phone_errors_failure_url if self.class != Idv::PhoneController
+      redirect_to idv_phone_errors_failure_url
+    when :proof_ssn
+      redirect_to idv_session_errors_ssn_failure_url
     end
   end
 
-  def throttle_and_controller_match(throttle_type)
-    case throttle_type
-    when :idv_resolution
-      self.instance_of?(Idv::VerifyInfoController) ||
-        self.instance_of?(Idv::InPerson::VerifyInfoController)
-    when :idv_doc_auth
-      self.instance_of?(Idv::DocumentCaptureController)
-    when :proof_address
-      self.instance_of?(Idv::PhoneController)
+  def idv_attempter_rate_limited?(throttle_type)
+    if throttle_type == :proof_ssn
+      return unless pii_ssn
+      Throttle.new(
+        target: Pii::Fingerprinter.fingerprint(pii_ssn),
+        throttle_type: :proof_ssn,
+      ).throttled?
+    else
+      Throttle.new(
+        user: idv_session_user,
+        throttle_type: throttle_type,
+      ).throttled?
     end
   end
 
-  def idv_attempter_rate_limited?(throttle_type)
-    Throttle.new(
-      user: idv_session_user,
-      throttle_type: throttle_type,
-    ).throttled?
+  def pii_ssn
+    return unless defined?(flow_session) && user_session
+    # pii_from_doc is stored in flow_session as either symbol or string
+    pii_from_doc_ssn = flow_session['pii_from_doc']&.[](:ssn)
+    pii_from_doc_ssn ||= flow_session[:pii_from_doc]&.[](:ssn)
```

Comment:
I wondered how you integrated :proof_ssn since it requires the SSN! Nice work with all the complexities here.

While the specs are picky about symbol or string, I think the user_session is defined with indifferent access, so you should be able to choose one or the other for the specs and go with that.